src/deluge/modulation/arpeggiator.cpp

```diff
@@ -128,47 +147,86 @@ void Arpeggiator::noteOn(ArpeggiatorSettings* settings, int32_t noteCode, int32_
 
 	lastVelocity = velocity;
 
-	int32_t n = notes.search(noteCode, GREATER_OR_EQUAL);
+	bool noteExists = false;
 
 	ArpNote* arpNote;
+	ArpNote* arpNoteAsPlayed;
 
-	// If it already exists...
-	if (n < notes.getNumElements()) {
-		arpNote = (ArpNote*)notes.getElementAddress(n);
+	int32_t notesKey = notes.search(noteCode, GREATER_OR_EQUAL);
+	if (notesKey < notes.getNumElements()) {
+		arpNote = (ArpNote*)notes.getElementAddress(notesKey);
 		if (arpNote->inputCharacteristics[util::to_underlying(MIDICharacteristic::NOTE)] == noteCode) {
-			if ((settings != nullptr) && settings->mode != ArpMode::OFF) {
-				return; // If we're an arpeggiator, return
-			}
-			else {
-				goto noteInserted; // Otherwise, try again. Actually is this really useful?
-			}
+			noteExists = true;
+		}
+	}
+	int32_t notesAsPlayedIndex = -1;
+	for (int32_t i = 0; i < notesAsPlayed.getNumElements(); i++) {
+		ArpNote* theArpNote = (ArpNote*)notesAsPlayed.getElementAddress(i);
+		if (theArpNote->inputCharacteristics[util::to_underlying(MIDICharacteristic::NOTE)] == noteCode) {
+			arpNoteAsPlayed = theArpNote;
+			notesAsPlayedIndex = i;
+			noteExists = true;
+			break;
 		}
 	}
 
-	// Insert
-	{
-		int32_t error = notes.insertAtIndex(n);
+	if (noteExists) {
+		// If note exists already, do nothing if we are an arpeggiator, and if not, go to noteinserted to update
+		// midiChannel
+		if ((settings != nullptr) && settings->mode != ArpMode::OFF) {
+			return; // If we're an arpeggiator, return
+		}
+		else {
+			goto noteInserted;
+		}
+	}
+	// If note does not exist yet in the arrays, we must insert it in both
+	else {
+		// Insert in notes array
+		int32_t error = notes.insertAtIndex(notesKey);
 		if (error) {
 			return;
 		}
-	}
-
-	arpNote = (ArpNote*)notes.getElementAddress(n);
-
-	arpNote->inputCharacteristics[util::to_underlying(MIDICharacteristic::NOTE)] = noteCode;
-	arpNote->velocity = velocity;
-	arpNote->outputMemberChannel =
-	    MIDI_CHANNEL_NONE; // MIDIInstrument might set this, but it needs to be MIDI_CHANNEL_NONE until then so it
-	                       // doesn't get included in the survey that will happen of existing output member channels.
+		// Save arpNote
+		arpNote = (ArpNote*)notes.getElementAddress(notesKey);
+		arpNote->inputCharacteristics[util::to_underlying(MIDICharacteristic::NOTE)] = noteCode;
+		arpNote->velocity = velocity;
+		arpNote->outputMemberChannel =
+		    MIDI_CHANNEL_NONE; // MIDIInstrument might set this, but it needs to be MIDI_CHANNEL_NONE until then so it
+		                       // doesn't get included in the survey that will happen of existing output member
+		                       // channels.
```

Comment:
style nit: move comment above assignment, since you're touching this line anyway.